Documentation/guides/D8andR8.md

```diff
@@ -0,0 +1,270 @@
+
+In other words, what is currently happening *before* we introduce D8/R8 support?
+
+1. The [Javac](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Javac.cs)
+   MSBuild task compiles `*.java` files to a `classes.zip` file.
+2. The [Desugar](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Desugar.cs)
+   MSBuild task "desugars" using `desugar_deploy.jar` if
+   `$(AndroidEnableDesugar)` is `True`.
+3. The [Proguard](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Proguard.cs)
+   MSBuild task shrinks the compiled Java code if
+   `$(AndroidEnableProguard)` is `True`. Developers may also supply
+   custom proguard configuration files via `ProguardConfiguration`
+   build items.
+4. The [CreateMultiDexMainDexClassList](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/CreateMultiDexMainDexClassList.cs)
+   MSBuild task runs `proguard` to generate a final, combined
+   `multidex.keep` file if `$(AndroidEnableMultiDex)` is `True`.
+   Developers can also supply custom `multidex.keep` files via
+   `MultiDexMainDexList` build items.
+5. The [CompileToDalvik](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs)
+   MSBuild task runs `dx.jar` to generate a final `classes.dex` file
+   in `$(IntermediateOutputPath)android\bin`. If `multidex` is
+   enabled, a `classes2.dex` (and potentially more) are also generated
+   in this location.
+
+# What will this process look like with D8 / R8?
+
+Two new MSBuild tasks named `R8` and `D8` will be created.
+
+1. The [Javac](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Javac.cs)
+   MSBuild task will remain unchanged.
+2. `D8` will run if `$(AndroidEnableMultiDex)` is `False`,
+   `$(AndroidLinkTool)` is not `r8`, and "desugar" by default.
+3. Otherwise, `R8` will run if `$(AndroidEnableMultiDex)` is `True` or
+   `$(AndroidLinkTool)` is `r8` and will also "desugar" by default.
+
+So in addition to be being faster in general (if Google's claims are
+true), we will be calling a *single* command line tool to produce dex
+files!
+
+# So how do developers use it? What are sensible MSBuild property defaults?
+
+Currently, a `csproj` file might have the following properties:
+```xml
+<Project>
+    <PropertyGroup>
+        <AndroidEnableProguard>True</AndroidEnableProguard>
+        <AndroidEnableMultiDex>True</AndroidEnableMultiDex>
+        <AndroidEnableDesugar>True</AndroidEnableDesugar>
+    </PropertyGroup>
+</Project>
+```
+
+To enable the new behavior, we should introduce two new enum-style
```

Comment:
Again, tense: at the time the PR is merged, this isn't describing *proposed* behavior.  This is documenting *actual* behavior:

> Use the `$(AndroidDexGenerator)` and `$(AndroidLinkTool)` properties to control behavior...